### PR TITLE
Make clear .Internal convention does not adhere to the spec

### PR DESCRIPTION
**Pull request by hasufell**

### CODE REVIEW

**Comment by gbaz** [comment]
I generally support this, though I think it could use some tweaks to the language to be a bit less abrupt, and also perhaps add a bit more motivation (it is a faq after all)!

File: pvp-faq.md [hunk 1]
```diff
@@ -291,3 +291,16 @@ conditionally, as you'd need to have access to the `Generic` class and
 be able to define instances for it in the first place, before being
 able to make use of (and thus observe) the default implementation of
 `rnf`.
+
+## Internal modules
+
+There has been a long standing convention in the haskell ecosystem to use
+modules that have a module name prefix/infix/suffix component named `Internal`
+(e.g. `Data.Text.Internal`) and then making these modules exempt from PVP.
```
Comment (ffaf1) [inline pvp-faq.md:299]: First paragraph should specify what `.Internal` are for (i.e. internal API).

Comment (hasufell) [reply pvp-faq.md:299]: Can you be a bit more verbose?

Comment (ffaf1) [reply pvp-faq.md:299]: The paragraph does not explain *why* `.Internal` modules are used. Suggestion:

> There is custom in the Haskell ecosystem to expose a library internal API via `.Internal` modules
(e.g. `Data.Text.Internal`) and then making these modules exempt from PVP.

My edit also remove the word “convention” which can mean two similar-yet-different things: a “general agreement”  or an “arbitrary custom”.

Comment (phadej) [reply pvp-faq.md:299]: FWIW, `Data.Text.Internal` **is not exempt** from PVP. It's is adhering to PVP and is exporting internals (for advanced usage).

In other words, `Data.Text.Internal` **is a bad example**.

Also the wording is implying that "`Internal` modules are exempt from PVP" practice is widely accepted, but **that is not true** either, as the discussion on #46 has indicated.

The wording here can cause more harm than clarification.

---

I'd try with something like:

> There is a dubious practice of exempting some modules from PVP. Various methods are used:
>
> * Mentioning in the documentation that the module is not under PVP
> * Naming module with prefix/infix/suffix component `Internal`.
> * Combination of above
>
> This practice and any others exemptions from the policy are not permitted by the PVP.
>
> The `Internal` modules are useful to export internals of a library, like in `Data.Text.Internal` in `text`.
>
> If author desires to keep library version churn to a minimum, yet be able to change internals freely,
> then they are adviced to create a separate package for the internal API.
> That approach is outlined  [in Nikita Volvov's blog post "Internal convention is a mistake"](http://nikita-volkov.github.io/internal-convention-is-a-mistake/).

**Comment by bgamari** [comment]
Thanks for carrying this out, @hasufell.

File: pvp-faq.md (hunk 1 above)
Comment (hasufell) [reply pvp-faq.md]: I adopted your wording mostly.

**Comment by tomjaguarpaw** [comment]
> I'm slightly against this addition to the spec ... I don't think a spec should mention practices that it forbids

This is not an addition to the *spec*. It's an addition to the FAQ.

**Comment by hasufell** [comment]
Exactly. This is an addition to the FAQ on a best-effort basis. We introduce this, because it's wide-spread practice (that's ok), but we want to raise awareness that it's not PVP conform and we want to promote the alternative (the `-internal` package approach). Adding this to PVP FAQ gives this better visibilty.

**Comment by chshersh** [comment]
Sure, if it's FAQ than I'm fine with the addition.

**Comment by parsonsmatt** [comment]
I would rather have seen #46 merged and approved. The `.Internal` pattern is all over the place and significantly easier for library maintainers. I do see the *value* of the `-internal` package convention, and I do agree that it is *better*, but it's also significantly more work.

Being totally 100% PVP compliant is a difficult task, and I'd be surprised if >1% of Hackage packages were totally compliant (noting that a properly compliant package would require a *minor bound* on `base` if you're implicitly or openly importing the `Prelude`). So, sure, this pattern isn't PVP compliant, might as well add it to the FAQ as an explicit call-out, but it's really just another demonstration that the PVP is an annoying policy and inflexible to the actual needs of users.

**Comment by hasufell** [comment]
> I would rather have seen #46 merged and approved.

That is entirely orthogonal to this issue (and I do not wish to discuss that here). Doing that would require a version increment of the spec.

The current versions of the spec do not permit that convention and there's no question about that. This PR just makes that more verbosely clear.

**Comment by Bodigrim** [comment]
@hasufell are you happy to trigger a vote? Or shall we wait for more opinions / discussion?

**Comment by hasufell** [comment]
I'm ok with triggering a vote.

**Comment by tomjaguarpaw** [comment]
+1 from me

**Comment by hasufell** [comment]
+1

**Comment by angerman** [comment]
While I'd much rather see proper annotations per binding for this. I agree that this is pretty much the community adopted convention. Thus until such a time where we have better mechanisms, this seems to just provide a bit more clarity around expectations, I guess.

Weak +1.

**Comment by Bodigrim** [comment]
+1 from me.

**Comment by Bodigrim** [comment]
@chshersh @mixphix @parsonsmatt just a gentle reminder to vote.

**Comment by chshersh** [comment]
+1

**Comment by mixphix** [comment]
-1. Should we expect all packages currently following the `.Internal` module convention to switch? If so, how soon? Surely we can't *force* such a change. And what would the effect be? That we stop using packages that adhere to the former? That dependents of internal functionality must import another package, trickling up through the ecosystem? The module-level demarcation is enough to indicate that functionality should be expected to break. Moving that to the package level just adds another obstacle for maintainers to navigate.

**Comment by parsonsmatt** [comment]
+1

The addition to the FAQ is technically correct, and the mention of the `-internal` pattern is a good step. 

I do agree with @mixphix that we should change the spec to match the common practice, but that seems to be an orthogonal concern.

**Pull request merged**